src/librustc_typeck/check/wfcheck.rs

```diff
@@ -251,16 +251,84 @@ impl<'ccx, 'gcx> CheckTypeWellFormedVisitor<'ccx, 'gcx> {
         });
     }
 
+    fn check_auto_trait(&mut self,
+                        trait_def_id: DefId,
+                        items: &[hir::TraitItem],
+                        span: Span)
+    {
+        // We want to ensure:
+        //
+        // 1) that there are no items contained within
+        // the trait defintion
+        //
+        // 2) that the definition doesn't violate the no-super trait rule
+        // for auto traits.
+        //
+        // 3) that the trait definition does not have any type parameters
+
+        let predicates = self.tcx().lookup_predicates(trait_def_id);
+
+        // We must exclude the Self : Trait predicate contained by all
+        // traits.
+        let has_predicates =
+            predicates.predicates.iter().any(|predicate| {
+                match predicate {
+                    &ty::Predicate::Trait(ref poly_trait_ref) => {
+                        let self_ty = poly_trait_ref.0.self_ty();
+                        !(self_ty.is_self() && poly_trait_ref.def_id() == trait_def_id)
+                    },
+                    _ => true,
+                }
+            });
+
+        let trait_def = self.tcx().lookup_trait_def(trait_def_id);
+
+        let has_ty_params =
+            trait_def.generics
+                      .types
+                      .len() > 1;
+
+        // We use an if-else here, since the generics will also trigger
+        // an extraneous error message when we find predicates like
+        // `T : Sized` for a trait like: `trait Magic<T>`.
+        //
+        // We also put the check on the number of items here,
+        // as it seems confusing to report an error about
+        // extraneous predicates created by things like
+        // an associated type inside the trait.
+
+        if !items.is_empty() {
+            error_380(self.ccx, span);
+        } else if has_ty_params {
+            span_err!(self.tcx().sess, span, E0566,
+                "traits with auto impls (`e.g. unsafe impl \
+                    Trait for ..`) can not have type parameters")
+        } else if has_predicates {
+            span_err!(self.tcx().sess, span, E0565,
+                "traits with auto impls (`e.g. unsafe impl \
+                    Trait for ..`) can not have predicates")
+        }
+
+        // Finally if either of the above conditions apply we should add a note
+        // indicating that this error is the result of a recent soundness fix.
+        if has_ty_params || has_predicates {
+            self.tcx().sess.span_note_without_error(
+                span,
+                "the new auto trait rules are the result of a \
+                    recent soundness fix; see #29859 for more details")
```

Comment:
Hmm, the best way to do this is actually to attach this to the original error I think. If you replace `span_err!` above with `struct_span_err!`, it will return an error object; you can store that into a variable, and then down here call `err.note(...)`, and -- at the end -- `err.emit()`